filesystems/blob/src/main/scala/cromwell/filesystems/blob/BlobPathBuilderFactory.scala

```diff
@@ -1,50 +1,115 @@
 import cromwell.core.path.PathBuilderFactory
 import net.ceedubs.ficus.Ficus._
 
-import java.time.OffsetDateTime
-import scala.concurrent.ExecutionContext
-import scala.concurrent.Future
+import java.net.URI
+import java.nio.file.{FileSystem, FileSystemNotFoundException, FileSystems}
+import java.time.temporal.ChronoUnit
+import java.time.{Duration, Instant, OffsetDateTime}
+import scala.concurrent.{ExecutionContext, Future}
 import scala.jdk.CollectionConverters._
+import scala.util.{Failure, Success, Try}
 
 final case class BlobFileSystemConfig(config: Config)
 final case class BlobPathBuilderFactory(globalConfig: Config, instanceConfig: Config, singletonConfig: BlobFileSystemConfig) extends PathBuilderFactory {
-  val container: String = instanceConfig.as[String]("store")
-  val endpoint: String = instanceConfig.as[String]("endpoint")
-  val workspaceId: Option[String] = instanceConfig.as[Option[String]]("workspace-id")
-  val workspaceManagerURL: Option[String] = singletonConfig.config.as[Option[String]]("workspace-manager-url")
+  val container: BlobContainerName = BlobContainerName(instanceConfig.as[String]("container"))
+  val endpoint: EndpointURL = EndpointURL(instanceConfig.as[String]("endpoint"))
+  val workspaceId: Option[WorkspaceId] = instanceConfig.as[Option[String]]("workspace-id").map(WorkspaceId(_))
+  val expiryBufferMinutes: Long = instanceConfig.as[Option[Long]]("expiry-buffer-minutes").getOrElse(10)
+  val workspaceManagerURL: Option[WorkspaceManagerURL] = singletonConfig.config.as[Option[String]]("workspace-manager-url").map(WorkspaceManagerURL(_))
 
   val blobTokenGenerator: BlobTokenGenerator = BlobTokenGenerator.createBlobTokenGenerator(
     container, endpoint, workspaceId, workspaceManagerURL)
+  val fsm: BlobFileSystemManager = BlobFileSystemManager(container, endpoint, expiryBufferMinutes, blobTokenGenerator)
 
   override def withOptions(options: WorkflowOptions)(implicit as: ActorSystem, ec: ExecutionContext): Future[BlobPathBuilder] = {
     Future {
-      new BlobPathBuilder(blobTokenGenerator, container, endpoint)
+      new BlobPathBuilder(container, endpoint)(fsm)
     }
   }
 }
 
-sealed trait BlobTokenGenerator {
-  def getAccessToken: String
+final case class BlobContainerName(value: String) {override def toString: String = value}
+final case class StorageAccountName(value: String) {override def toString: String = value}
+final case class EndpointURL(value: String) {override def toString: String = value}
+final case class WorkspaceId(value: String) {override def toString: String = value}
+final case class WorkspaceManagerURL(value: String) {override def toString: String = value}
+
+object BlobFileSystemManager {
+  def parseTokenExpiry(token: AzureSasCredential): Option[Instant] = for {
+    expiryString <- token.getSignature.split("&").find(_.startsWith("se")).map(_.replaceFirst("se=","")).map(_.replace("%3A", ":"))
+    instant = Instant.parse(expiryString)
+  } yield instant
```

Comment:
Hmmm, do we expect to ever return None from this method? It looks like that would result in us getting a new token and refreshing the filesystem every time we need it. For now we should perhaps throw an error instead. I'd rather it blow up than silently refresh itself constantly.